internal/controllers/machine/machine_controller_inplace_update.go

```diff
@@ -0,0 +1,266 @@
+		return ctrl.Result{}, nil
+	}
+	if !ptr.Deref(s.machine.Status.Initialization.BootstrapDataSecretCreated, false) {
+		log.V(5).Info("Bootstrap data secret not yet created, skipping in-place update")
+		return ctrl.Result{}, nil
+	}
+
+	if s.infraMachine == nil {
+		s.updatingReason = clusterv1.MachineUpdateFailedReason
+		s.updatingMessage = "In-place update not possible: InfraMachine not found"
+		return ctrl.Result{}, errors.New("in-place update failed: InfraMachine not found")
+	}
+
+	infraReady := r.isInfraMachineReadyForUpdate(s)
+	bootstrapReady := r.isBootstrapConfigReadyForUpdate(s)
+
+	if !infraReady || !bootstrapReady {
+		log.Info("Waiting for InfraMachine and BootstrapConfig to be marked for in-place update")
+		return ctrl.Result{}, nil
+	}
+
+	log.Info("UpdateMachine hook is pending, calling runtime hook")
+	result, message, err := r.callUpdateMachineHook(ctx, s)
+	if err != nil {
+		s.updatingReason = clusterv1.MachineUpdateFailedReason
+		s.updatingMessage = fmt.Sprintf("UpdateMachine hook failed: %v", err)
+		return ctrl.Result{}, err
+	}
+
+	if result.RequeueAfter > 0 {
+		s.updatingReason = clusterv1.MachineWaitingForUpdateMachineHookReason
+		if message != "" {
+			s.updatingMessage = fmt.Sprintf("UpdateMachine hook in progress: %s", message)
+		} else {
+			s.updatingMessage = "UpdateMachine hook in progress"
+		}
+		return result, nil
+	}
+
+	log.Info("In-place update completed successfully")
+	if err := r.completeInPlaceUpdate(ctx, s); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to complete in-place update")
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// isInfraMachineReadyForUpdate checks if the InfraMachine has the in-place update annotation.
+func (r *Reconciler) isInfraMachineReadyForUpdate(s *scope) bool {
+	_, hasAnnotation := s.infraMachine.GetAnnotations()[clusterv1.UpdateInProgressAnnotation]
+	return hasAnnotation
+}
+
+// isBootstrapConfigReadyForUpdate checks if the BootstrapConfig has the in-place update annotation.
+func (r *Reconciler) isBootstrapConfigReadyForUpdate(s *scope) bool {
+	if s.bootstrapConfig == nil {
+		return true
+	}
+	_, hasAnnotation := s.bootstrapConfig.GetAnnotations()[clusterv1.UpdateInProgressAnnotation]
+	return hasAnnotation
+}
+
+// callUpdateMachineHook calls the UpdateMachine runtime hook for the machine.
+func (r *Reconciler) callUpdateMachineHook(ctx context.Context, s *scope) (ctrl.Result, string, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Validate that exactly one extension is registered for the UpdateMachine hook.
+	// For the current iteration, we only support a single extension to ensure safe behavior.
+	// Support for multiple extensions will be introduced in a future iteration.
+	extensions, err := r.RuntimeClient.GetAllExtensions(ctx, runtimehooksv1.UpdateMachine, s.machine)
+	if err != nil {
+		return ctrl.Result{}, "", errors.Wrap(err, "failed to get registered UpdateMachine extensions")
```

Comment:
```suggestion
		return ctrl.Result{}, "", err
```
I think the error build inside of GetAllExtensions already has enough context, this would just repeat it a bit